### ISSUE TITLE
[Bug] Logging queries outputs binary objectguid, causes SQL error with Telescope and MS SQL Server

### ISSUE THREAD

**Issue by Spice-King** [DirectoryTree/LdapRecord#293]
**Environment:**
 - LDAP Server Type: ActiveDirectory
 - LdapRecord-Laravel Major Version: v2
 - PHP Version: 8

This might be more Laravel tied, but I've traced the source of this to here when I search by GUID.
https://github.com/DirectoryTree/LdapRecord/blob/13879553fed736db3d5225662337cb99ec5662ba/src/Events/Logger.php#L110-L128

For most things, un-escaped is probably fine. However, a functionally random bit of binary (the raw objectguid) thrown into a string is causing Telescope -> Laravel -> Microsoft SQL Server chain to buck, hard. The content Telescope wanted to store, an object which gets converted to a JSON string, gets converted to `0` and, SQL Server being SQL Server, thinks that all entries for that column should be converted into `int`s in spite of it being a `nvarchar` column. The `0` is a converted `false` from PHP failing to convert the object to a JSON string, error from that is 'Malformed UTF-8 characters, possibly incorrectly encoded'.

This has a few knock on effects, Telescope's batch insert fails for it's event collection, which eats all events generated for that request in Telescope, logging is included as an event here. I still get basic logging, files in my case since we rarely need them, but the SQL exception it generated results in about 1/4 MB of SQL code being dumped out on a single line (the result of enumerating all groups a user has in AD, plus throwing it into cache). As you can assume, many text editors really don't cope well with ~1/4 MB of text on a single line.

Disabling logging for LdapRecord will be a thing once this gets into production, which would bypass this, but I kinda need that on for developing and testing during our migration from using Adldap2. I'm sure that there are other logging systems that do not handle binary being in a string well either, as it's not expected to be in stream of human readable logs.

I don't know what would be a good solution that does not worm its way through everything to fix this. Perhaps wrapping binary values in a class that can translate it into a usable non-binary result. One could make specialized wrappers for a few known types like the GUIDs, but I'd be happy with just converting it to hex or something. A simpler option would be to replace every non ASCII character when logging with a marker and a hex encoded version of it, like \xFF, or something, but that will clobber most Unicode symbols despite being valid and can result in flipping back and forth with the escaping the binary value.

**Comment by stevebauman** [comment]
Thanks for the extremely detailed issue @Spice-King! ❤️ 

Let's get this fixed.

I think we could leverage the event itself to handle converting the characters of the unescaped query, for example:

```php
// ...

$message = "..."
    ." - Filter: {$event->getLoggableQuery()}";

// ...
```

Though, the unescaped query should contain a hexadecimal version of an object GUID (if you're using `Model::find($stringGuid)`. Is this happening with `Model::find()`? Or is this happening with a raw/custom query?

**Comment by Spice-King** [comment]
I'm getting this by doing `Model::query()->findByGuidOrFail($stringGuid)`. (Phpstorm complains if I omit query(), since it thinks findByGuidOrFail is not static, and I'm more than happy to do it the longer way.)

`Model::find($stringGuid)` errors with `ldap_read(): Search: Invalid DN syntax`, which makes sense, since it's just a GUID and it's expecting a DN.

`HasLdapUser` from the Laravel package also seems to cause this as well, but I have no plans to use that. Does not work for checking stuff on other users or via the console for jobs and tasks.

**Comment by stevebauman** [comment]
> Model::find($stringGuid) errors with ldap_read(): Search: Invalid DN syntax, which makes sense, since it's just a GUID and it's expecting a DN.

Ah right, my mistake -- I definitely meant `Model::findByGuid()` 😅 

Are you able to see if this happens if you modify the `Logger` inside of your `vendor` directory and use `$query->getQuery()` instead of `$query->getUnescapedQuery()`?

I'm expecting it to fail, since the escaped query will contain hex. Let me know!

**Comment by Spice-King** [comment]
Worked fine! Did dump out everything like `(objectclass=\74\6f\70)(objectclass=\70\65\72\73\6f\6e)(objectclass=\6f\72\67\61\6e\69\7a\61\74\69\6f\6e\61\6c\70\65\72\73\6f\6e)(objectclass=\75\73\65\72)objectclass=\63\6f\6d\70\75\74\65\72)` as apossed to `(objectclass=top)(objectclass=person)(objectclass=organizationalperson)(objectclass=user)`. No binary output, debugbar happy, and Telescope logging away with out exploading.

PHP's hex escape is `\x##`, and that only applies to text literals, I think. Dn't think you can trigger them by using regex to replace contents of a string.

**Comment by stevebauman** [comment]
Oh strange! Okay, well that's good that it works, since we can use that as worst case.

Do you know if it's possible to easily escape those binary sequences? I'd like to try and mantain the ability to see the raw unescaped query, so it's easier to debug for developers.

**Comment by Spice-King** [comment]
Going off this [ASCII chart](http://en.cppreference.com/w/cpp/language/ascii) 32 to 126 are safe to print. Excluding the backslash (92) and possibly space (32) too and allowing the rest in the range would create an unambitious escape with allowing almost all of the normal characters through. Most web devs, or experienced web users even, should recognize `0x20` as space from URL encoding, so that should be fine to exclude. Though this is totally ignoring anything Unicode, which may be for the best since invalid code points could be generated.

https://github.com/DirectoryTree/LdapRecord/blob/13879553fed736db3d5225662337cb99ec5662ba/src/Utilities.php#L46-L51

```php
public static function unescape($value)
{
  return preg_replace_callback('/\\\([0-9A-Fa-f]{2})/', function ($matches) {
    $v = hexdec($matches[1]);
    if ($v > 31 && $v < 127 && $v != 92) return chr($v);
    return $match[0];
  }, $value);
}
```

This is a quick and dirty implementation written in the comment box, then figured I'd test it by throwing it in Tinker.
```
>>> $f='\\11\\f8\\5c\\73\\7f\\f4\\bc\\41\\46\\ff\\e7\\92\\d0\\73\\f4\\1f'
=> "\11\f8\5c\73\7f\f4\bc\41\46\ff\e7\92\d0\73\f4\1f"
>>> LdapRecord\Utilities::unescape($f)
=> "\11\f8\5cs\7f\f4\bcAF\ff\e7\92\d0s\f4\1f"
>>>
```

**Comment by stevebauman** [comment]
Thanks @Spice-King! I've decided to do the easiest fix -- by simply only logging the escaped query. This will allow devs to copy-and-paste the filter in their own LDAP browser to debug, without having to splice out the GUID.

I played with manually escaping GUIDs in the filter, but it printed characters that broke the filter (included brackets, for example).

I've just released [v2.4.8](https://github.com/DirectoryTree/LdapRecord/releases/tag/v2.4.8) with this patch.

Thanks again for your help and detailed report! 👍
